source/common/upstream/upstream_impl.h

```diff
@@ -69,13 +70,20 @@ class HostDescriptionImpl : virtual public HostDescription {
         canary_(Config::Metadata::metadataValue(metadata, Config::MetadataFilters::get().ENVOY_LB,
                                                 Config::MetadataEnvoyLbKeys::get().CANARY)
                     .bool_value()),
-        metadata_(metadata), locality_(locality), stats_{ALL_HOST_STATS(POOL_COUNTER(stats_store_),
-                                                                        POOL_GAUGE(stats_store_))} {
-  }
+        metadata_(std::make_shared<envoy::api::v2::core::Metadata>(metadata)),
+        locality_(locality), stats_{ALL_HOST_STATS(POOL_COUNTER(stats_store_),
+                                                   POOL_GAUGE(stats_store_))} {}
 
   // Upstream::HostDescription
   bool canary() const override { return canary_; }
-  const envoy::api::v2::core::Metadata& metadata() const override { return metadata_; }
+  void canary(bool is_canary) override { canary_ = is_canary; }
+  const std::shared_ptr<envoy::api::v2::core::Metadata> metadata() const override {
+    return metadata_;
```

Comment:
This not thread safe on read, since shared_ptr needs to both read the value and increment the ref count. This will need to be guarded with a read lock on the shared mutex. Two additional things:
1) Can you look into adding lock annotations for this lock? IIRC this doesn't work correctly for reader/writer locks last I checked so it might not be possible, but I know @jmarantz did some work in this area so maybe it is now? Also, are we supposed to use absl locks now? Check https://github.com/envoyproxy/envoy/blob/master/source/common/common/thread.h. If there is no R/W lock there now just add a TODO comment here about adding that later.

2) Can you add some comment here about perf? If lock contentions becomes an issue we can refactor later to use TLS and posting from the main thread.

Reply:
Yup — will do. Thanks for the review!